Adding a design proposal template and README. (#1838)
GoogleContainerTools/skaffold: docs/design_proposals/design-proposal-template.md

# Title

* Author(s): \<your name\>
* Design Shepherd: \<skaffold-core-team-member\>

    If you are already working with someone mention their name.
    If not, please leave this empty, it will be assigned to a core team member.
* Date: \<date\>
* Status: [Draft/Reviewed/Complete]

## Background

In this section, please mention and describe the new feature, redesign
or refactor.

Please provide a brief explanation for the following questions:

1. Why is this required?
2. If this is a redesign, what are the drawbacks of the current implementation?
3. Is there any another work-around, and if so, what are its drawbacks?
4. Mention related issues, if there are any.

Here is an example snippet for a new feature:

___
Currently, skaffold config supports `artifact.sync` as a way to sync files
directly to pods. So far, artifact sync requires a specification of sync
patterns like

```yaml
sync:
  '*.js': app/
```

This is error prone and unnecessarily hard to use, because the destination is

> already contained in the Dockerfile for docker build. (see #1166, #1581).

Reviewer: no need to reproduce my typos :)
```suggestion
already contained in the Dockerfile for docker build (see #1166, #1581).
```

In addition, the syncing needs to handle special cases for globbing and often

> requires a long list of sync patterns (#1807)

Reviewer: ```suggestion
requires a long list of sync patterns (#1807).
```

___

## Design

Please describe your solution. Please list any:

* new config changes
* interface changes
* design assumptions

For a new config change, please mention:

* Is it backwards compatible? If not, what is the deprecation policy?
  Refer to the [deprecation policy requirements.](./../../deprecation-policy.md#how-do-we-deprecate-things)

>   for details)

Reviewer: ```suggestion
  for details.
```

### Open Issues/Question

Please list any open questions here in the format.

**\<Question\>**

Resolution: Please list the resolution if resolved during the design process or
specify __Not Yet Resolved__

## Implementation plan
As a team, we've noticed that larger PRs can go unreviewed for long periods of

> time. Small incremental changes get reviewed faster and also easier for reviewers.

Reviewer: ```suggestion
time. Small incremental changes get reviewed faster and are also easier for reviewers.
```

# Title

* Author(s): \<your name\>

> * Design Shepherd: As mentioned in the document here.

Reviewer: Maybe make this more self-explanatory?
```suggestion
* Design Shepherd: Please leave this empty, it will be assigned to a core team member.
```

    If you are already working with someone mention their name.
    If not, please leave this empty, it will be assigned to a core team member.
* Date: \<date\>

> * Status: [Draft/Reviewed/Complete]

Reviewer 2: I think Status should be Reviewed / Under implementation / Complete / Cancelled - as the Draft phase is implicit while it's in the PR phase. We should also explain what these are ,e.g.:

`Reviewed` means that the proposal PR has been merged 
`Under implementation` means that the plans is being executed by actual work
`Complete` means that there is no changes expected in this design anymore, if 
`Cancelled` means that we learned something during or before implementation that caused this proposal to be cancelled. 

I would add a `Links` section as well, to collect relevant PRs and issues too. 

What do you think?

Author: > I would add a Links section as well, to collect relevant PRs and issues too.

I think we have mentioned that in the `Background` bullet 4.

Author: > I think Status should be Reviewed / Under implementation / Complete / Cancelled - as the Draft phase is implicit while it's in the PR phase. We should also explain what these are ,e.g.:

So @balopat along your lines but a little different, 

1. In Review: The proposal PR has not been reviewed yet by the core team.
2. Accepted: The proposal PR has been accepted and ready for implementation.
3. Not Accepted: The proposal PR was not accepted by the core team.
4. Under implementation: An accepted proposal is being implemented by actual work.
5. Cancelled: During or before implementation the proposal was cancelled.
   It could be due to:

   * other features added which made the current design proposal obselete.
   * no longer a priority.

6. Complete: This feature/change is implemented.

or re-factor.

Please provide an rationale covering following points:

1. Why is this required?

> 2. If its re-design, What are cons with current implementation?

Reviewer: ```suggestion
2. If it is a re-design, what are the cons with the current implementation?
```

> 3. Are there any another work-around and if yes, why not keep using it.

Reviewer: ```suggestion
3. Are there any another work-arounds and if yes, why not keep using it.
```

> 

Reviewer: Maybe add an additional item
```
4. Mention related issues, if there are any.
```

1. Why is this required?
2. If this is a redesign, what are the drawbacks of the current implementation?

> 3. Is there any another work-around, and if so, what are its drawbacks?

Reviewer 2: ```suggestion
3. Is there any another workaround, and if so, what are its drawbacks?
```

Currently, skaffold config supports `artifact.sync` as a way to sync files
directly to pods. So far, artifact sync requires a specification of sync
patterns like

```yaml
sync:
  '*.js': app/
```

This is error prone and unnecessarily hard to use, because the destination is
already contained in the Dockerfile for docker build. (see #1166, #1581).
In addition, the syncing needs to handle special cases for globbing and often
requires a long list of sync patterns (#1807)
___

## Design

Please describe your solution. Please list any:

* new config changes
* interface changes
* design assumptions

For a new config change, please mention:

* If its a backward compatible config change

> * If its a backward compatible config change, is there a migration path?

Reviewer: I do not fully understand what is meant by the migration path. I presume, you do not mean the automatic schema upgrade? Can you elaborate on this one a bit more?

Author: thanks. i changed to mention how you are going to deprecate it with links to deprecation policy docs.

### Open Issues/Question

Please list any open questions here in the format.

**\<Question\>**

Resolution: Please list the resolution if resolved during the design process or
specify __Not Yet Resolved__

## Implementation plan
We have identified, huge PRs go unnoticed for a long time. Small incremental
changes get reviewed faster and also easier for reviewers.

> For a design feature, list a summary of tasks breakdown for e.g.:

Reviewer: I don't think that the differentiation between `design feature` and `re-factor proposal` is relevant here. The bottom line is to provide a summary of smaller tasks which should be submitted as separate pull requests, correct? This applies to both cases, such that making a distinction will only confuse people.

Author: okies i can remove this. :)

For the example desing proposal to infer artifact sync, some of the smaller task
could be:
___

1. Add new config key `infer` to `artifact.sync` and test schema validation.
2. Add inference logic for docker and examples.
3. Support both `infer` and user defined map with precedence rules implemented.

> 4. Finally, support builder plugins to add sync patterns.

Reviewer: Please also include along the lines
```
5. Adapt user documentation.
```
Because it is easily forgotten :)

# Title

* Author(s): \<your name\>
* Design Shepherd: \<skaffold-core-team-member\>

    If you are already working with someone mention their name.
    If not, please leave this empty, it will be assigned to a core team member.
* Date: \<date\>
* Status: [Draft/Reviewed/Complete]

## Background

> In this section, please mention and describe the new feature, re-design

Reviewer 3: s/`re-design`/`redesign`

s/`re-factor`/`refactor`

or re-factor.

> Please provide an rationale covering following points:

Reviewer 3: `Please provide a brief explanation for the following questions: `

1. Why is this required?

> 2. If its re-design, What are cons with current implementation?

Reviewer 3: `If this is a redesign, what are the drawbacks of the current implementation?`

> 3. Is there any another work-around and if yes, why not keep using it.

Reviewer 3: `Is there a known workaround, and if so, what are its drawbacks?`

4. Mention related issues, if there are any.

Here is an example snippet for a new feature:

___
Currently, skaffold config supports `artifact.sync` as a way to sync files
directly to pods. So far, artifact sync requires a specification of sync
patterns like

```yaml
sync:
  '*.js': app/
```

This is error prone and unnecessarily hard to use, because the destination is
already contained in the Dockerfile for docker build. (see #1166, #1581).
In addition, the syncing needs to handle special cases for globbing and often
requires a long list of sync patterns (#1807)
___

## Design

Please describe your solution. Please list any:

* new config changes
* interface changes
* design assumptions

For a new config change, please mention:

> * If its a backward compatible config change ?

Reviewer 3: let's move this bullet point into the previous question:

```
If there is a new config change, is it backwards compatible? If not, what is the deprecation policy?
Refer to the [deprecation policy requirements] ./../../deprecation-policy.md#how-do-we-deprecate-things) for details.
```

* If the answer to above question is yes, what would be the deprecation policy?
  See [deprecation-policy](./../../deprecation-policy.md#how-do-we-deprecate-things)
  requirements.

### Open Issues/Question

Please list any open questions here in the format.

**\<Question\>**

Resolution: Please list the resolution if resolved during the design process or
specify __Not Yet Resolved__

## Implementation plan

> We have identified, huge PRs go unnoticed for a long time. Small incremental

Reviewer 3: `As a team, we've noticed that larger PRs can go unreviewed for long periods of time.`

changes get reviewed faster and also easier for reviewers.

For a design feature, list a summary of tasks breakdown for e.g.:

> For the example desing proposal to infer artifact sync, some of the smaller task

Reviewer 3: s/`desing`/`design`

`For the example artifact sync proposal, some of the smaller tasks could be:`

could be:
___

1. Add new config key `infer` to `artifact.sync` and test schema validation.
2. Add inference logic for docker and examples.
3. Support both `infer` and user defined map with precedence rules implemented.
4. Finally, support builder plugins to add sync patterns.

___

## Integration test plan

> Please describe what new test cases are you going to consider.

Reviewer 3: s/`are you`/`you are`

1. Why is this required?
2. If its re-design, What are cons with current implementation?
3. Is there any another work-around and if yes, why not keep using it.

> 4. Mention related issues, if there are any.****

Reviewer: If the `****` is intentional, I don't get what it means

Author: no its unintentional.
